Added more functionalities to OptimizedRNN operator

yuanbyu opened the pull request:
Added two new cell types: 
 - lstm-peepholes: Do peephole in LSTM. Peephole weight is an additional input.
 - lstm-input-forget: Couple the input and forget gates.

Added the following attributes:
 - activations: A list of activation functions for the gates, so we could customize the activations.
 - use_bias: Determine whether to use the bias tensor. Default to true.
 - reverse: Reverse the inputs.  Default is false.
 - clip: Cell clip threshold. Default to no clip.

The weight packing in input 0 seems to be under-specified. For example, in the lstm case, it is unclear how Wi, Wf, Wo, and Wc are packed in the weight tensor. I can try to improve it if I know the layout.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -108,18 +136,19 @@ The order of matrixes `{K, L, D, R, N, C}` is defined as:
            "Optional initial value of the hidden. If not specified - assumed "
-           "to be 0. Dimensions `[num_layers * directions, batch_size, "
-           "hidden_size]`")
+           "to be 0. Dimensions `[num_layers * directions, hidden_size]`")
```
actually, why did you remove `batch_size`? Cudnn allows to initialize different state per example and it in general makes sense (if you want to continue previous sequence)

yuanbyu replied:
I don't know it. Reverted.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -108,18 +136,19 @@ The order of matrixes `{K, L, D, R, N, C}` is defined as:
+    .Input(4, "seq_lens",
+           "Optional tensor specifying lengths of the sequences in a batch."
+           "Has shape `[batch_size]`.")
+    .Input(5, "peephole_w",
```
shouldn't we merge it into the global block of weights above?

yuanbyu replied:
This is related to the layout of the weights: We may have to pad if the layout is not right. Also, it is optional. So I made it a separate input.  It is fine with me to pack it into weights.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -71,6 +90,15 @@ allows to make spec more uniform.
     .Attr("num_layers", "Numbers of RNN layers in the stack, default 1",
           AttrType::INT)
     .Attr("hidden_size", "Number of neurons in the hidden layer", AttrType::INT)
+    .Attr("activations", "A list of activation functions for the gates. Typical activation "
```
you need to describe the order of activation functions (probably the same as the gates order in weights blob below). And say that default version just assumes stuff from the formula above.

yuanbyu replied:
Done.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -71,6 +90,15 @@ allows to make spec more uniform.
+    .Attr("activations", "A list of activation functions for the gates. Typical activation "
+          "functions are sigmoid and tanh",
+          AttrType::STRINGS)
+    .Attr("use_bias", "Use the bias tensors for the gates if 1. Default 1.",
```
edit the comment for weights to describe that bias is optional in this case?

yuanbyu replied:
Biases are still included in input(0) but not used if use_bias is false.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -71,6 +90,15 @@ allows to make spec more uniform.
+          AttrType::INT)
+    .Attr("reverse", "Process the sequences in reverse order, default 0",
+          AttrType::INT)
+    .Attr("clip", "Cell clip threshold. Default to no clip if not specified.",
```
add more comments in the spec above regarding where exactly 'clip' gets plugged in?

yuanbyu replied:
Done.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -82,7 +110,7 @@ Each parameter matrix is linearly appended after the previous parameter matrix
 without padding.
 
 The order of matrixes `{K, L, D, R, N, C}` is defined as:
```
you'd need to update the comment here too - for example change 'relu' / 'tanh' to 'simple' and update layout order

yuanbyu replied:
Thanks. Done.
Regarding the layout order in the weights input, I don't have a complete picture. Take a simple example, for LSTM, it is not completely clear to me how we pack Wi, Wf, Wo, and Wc in the weight tensor. Is it [4*hidden_size, input_size] or [hidden_size, 4*input_size]?  Perhaps I missed something obvious.

dzhulgakov replied:
It's actually more complicated as we're trying to pack several layers of e.g. LSTM together. The first layer would have different input size from later ones, the input gates might not have inputs at all, etc. Right now I just concatenated all the weight matrices together. It's not pretty but solves the problem of too many inputs.

Thus for peep_hole connection I'd just modify the message to say that there are more weight matrices in the internal order (R being "input-hidden", "hidden-hidden" and "peephole").

yuanbyu replied:
@dzhulgakov As we discussed, I will try to update this PR to have 4 RNN ops: SimpleRNN, GRU, LSTM, and BirectionalLSTM. I think that this would make things simpler.

dzhulgakov replied:
Yes, adding separate ops would make it simpler. The only problem is that stacked nature - for example the input-hidden matrix has different sizes for the first and subsequent layers.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -108,18 +136,19 @@ The order of matrixes `{K, L, D, R, N, C}` is defined as:
            "specified - assumed to be 0. Dimensions `[num_layers * directions, "
-           "batch_size, hidden_size]`")
+           "hidden_size]`")
+    .Input(4, "seq_lens",
```
with several optional args the ordering is unfortunate. Maybe we should allow for 'None' (or empty string) in inputs for this purpose

yuanbyu replied:
Reverted.

raduk commented on docs/Operators.md:
```diff
@@ -885,47 +887,72 @@ Equation definitions:
+
+  `lstm_input_forget with default activations`
+  - it = sigmoid(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
+  - ft = sigmoid(Wf*Xt + Rf*Ht-1 + Wbf + Rbf)
```
This is not needed anymore. We may also comment that forget weights are ignored for this type of cell

yuanbyu replied:
Good catch. Removed.

raduk commented on docs/Operators.md:
```diff
@@ -885,47 +887,72 @@ Equation definitions:
 - Equations:
-  `relu`
+  `simple(relu)`
   - Ht = ReLU(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
```
We may add the clip to all the formulas: ReLU(clip(Wi*Xt + Ri*Ht-1 + Wbi + Rbi))

yuanbyu replied:
Clip is optional and disabled as default. There are comments on the clip attribute.

yuanbyu commented:
@dzhulgakov As I understand it, this one is blocked due to the optional inputs issue #151?

dzhulgakov commented:
Yep, let me hack #151.

In the meantime we also need to figure out what to do with peephole connections weights. If you have better suggestions on how to layout the weight matrices - please share.

yuanbyu commented:
@dzhulgakov @raduk @ebarsoum I have updated this PR as we discussed:

1. Removed OptimizedRNN and replaced it with 4 simpler ops: SimpleRNN, GRU, LSTM, BidirectionalLSTM. 
2. The current version doesn't support multi-layer RNNs. We will add it in a separate PR with a new version number.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+          AttrType::INT)
+    .Attr("input_forget", "Couple the input and forget gates if 1, default 0.",
+          AttrType::INT)	  
+    .Attr("clip", "Cell clip threshold. Clipping bounds the elements of a tensor "
```
for repetitive arguments like this you may find FillUsing useful and extract common args to a subfunction. See for example https://github.com/onnx/onnx/blob/ea6d6578d9d2e566be69077a8c5b43663b0649c7/onnx/defs/reduction/defs.cc#L36

yuanbyu replied:
Good point. Done.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+	  "and tanh. See the equations for default if not specified.",
+          AttrType::STRINGS)
+    .Attr("hidden_size", "Number of neurons in the hidden layer", AttrType::INT)
+    .Attr("input_forget", "Couple the input and forget gates if 1, default 0.",
```
so if you couple them the input weight would be different? should we instead create a yet another op instead?

yuanbyu replied:
My understanding is W_i is ignored in this case. I am inclined not to create a new op for it.

raduk replied:
I'd add this: the forget gate is computed as f = 1-i. The weights for the forget gate are ignored.

yuanbyu commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+Equations:
+  - Ht = Activation(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
+)DOC")
+    .Attr("activation", "The activation function for input gate. Typical "
```
Thanks. Done.

yuanbyu commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+    .Input(0, "input",
+           "The input sequences packed (and potentially padded) into one 3-D "
+           "tensor with the shape of `[seq_length, batch_size, input_size]`.")
+    .Input(1, "W", 
```
I would prefer to keep it not optional for this PR, and do it in the next one.

yuanbyu commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+	   "`Wbi` and `Rbi`, and has shape `[2*hidden_size]`, Optional: If not "
+	   "specified - assumed to be 0.",
+	   true /*optional*/)
+    .Input(4, "initial_h",
```
I have a slight preference to keep it simple as is.

yuanbyu commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+    .Input(1, "W",
+	   "The weight tensor for the gates. The weights W[zrh] are "
+	   "concatenated along dimension 0. This tensor has shape "
+	   "`[3 * hidden_size, input_size]`.")
```
It seems that both CoreML and PyTorch use this notation so I will keep it as is.

yuanbyu commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+    .Input(3, "B",
+	   "The bias tensor for the gates. The biases Wb[zrh] and Rb[zrh] are "
+	   "concatenated along dimension 0. This tensor has shape "
+	   "`[6 * hidden_size]`. Optional: If not specified - assumed to be 0",
```
See above.

yuanbyu commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
-  `lstm`
+`H` - Hidden state
+
+Equations (LSTM with default activations):
```
Added the peephole term in the equation.

yuanbyu commented on onnx/defs/rnn/defs.cc:
```diff
@@ -5,124 +5,311 @@
+	    "`[batch_size, hidden_size]`.");
+
+
+OPERATOR_SCHEMA(BidirectionalLSTM)
```
Changed to have bidirectional support for all 3 ops.

yuanbyu commented:
@dzhulgakov I made a round of edits based on the comments. PTAL. Thanks.

raduk commented on onnx/defs/rnn/defs.cc:
```diff
@@ -159,158 +164,56 @@ custom implementation such as CuDNN.
 `Wb[iofc]` - W bias vectors for input, output, forget, and cell gates
 `Rb[iofc]` - R bias vectors for input, output, forget, and cell gates
 `P[iof]`  - P peephole weight matrix for input, output, and forget gates
```
shouldn't this be a vector ? same shape as R ?

raduk commented on onnx/defs/rnn/defs.cc:
```diff
@@ -6,6 +6,37 @@
+        schema.Attr("clip", "Cell clip threshold. Clipping bounds the elements of a tensor "
+                    "in the range of [-threshold, +threshold] and is applied to the input "
+                    "of activations. No clip if not specified.",
+                    AttrType::FLOAT);
```
Coming back to this: it feels a bit weird to have the default defined by the absence. And it is not consistent with other attributes that have defaults.

raduk replied:
Also, from the description it may not be clear where this is applied (I've seen people mistakenly clipping after the activation for instance, or only on parts of the equation). I'd add the clip in the equations.

raduk commented on onnx/defs/rnn/defs.cc:
```diff
@@ -159,158 +164,56 @@ custom implementation such as CuDNN.
+    .Input(7, "P",
+	   "The weight tensor for peepholes. Concatenation of `P[iof]` and "
+	   "`PB[iof]` (if bidirectional) along dimension 0. It has shape "
+	   "`[num_directions, 3*hidde_size, hidden_size]`. Optional: If not specified - "
```
[num_directions, 3*hidden_size] ?

yuanbyu replied:
Yes. Thanks.

yuanbyu commented on docs/Operators.md:
```diff
@@ -845,6 +993,82 @@
+  `R[iofc]` - R recurrence weight matrix for input, output, forget, and cell gates
+  `Wb[iofc]` - W bias vectors for input, output, forget, and cell gates
+  `Rb[iofc]` - R bias vectors for input, output, forget, and cell gates
+  `P[iof]`  - P peephole weight matrix for input, output, and forget gates
```
Good point. Fixed.

raduk commented on onnx/defs/rnn/defs.cc:
```diff
@@ -6,125 +6,214 @@
+Equations:
+  - Ht = Activation(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
+)DOC")
+    .Attr("activation", "The activation function for input gate. It must be "
```
If bidirectional we need more of these

yuanbyu replied:
Good point. I will update the spec.

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -6,125 +6,206 @@
+	   "`[num_directions, 3*hidde_size]`. Optional: If not specified - "
+	   "assumed to be 0.", "T",
+	   true /*optional*/)
+    .FillUsing(RNNDocGenerator("LSTM"));
```
sorry for nitpicking, but the order of arguments is weird in this case:

X,W,R,B
initial_h
seq_lengths
initial_c
P

Maybe group them together? At least seq_len shouldn't separate out init_c and init_h. As for P it's fine as it's rarely used, so we can keep it last

dzhulgakov replied:
@yuanbyu - would you mind to address this comment? 'Conditional accept' assumes that issues are addressed before merging :)

yuanbyu replied:
Sure. I will have a small PR for this coming your way. :-)

dzhulgakov commented on onnx/defs/rnn/defs.cc:
```diff
@@ -6,125 +6,206 @@
+`RB[iofc]` - R recurrence weight matrix for backward input, output, forget, and cell gates
+`WBb[iofc]` - W bias vectors for backward input, output, forget, and cell gates
+`RBb[iofc]` - R bias vectors for backward input, output, forget, and cell gates
+`PB[iof]`  - P peephole weight vector for backward input, output, and forget gates
```
denote that it's optional

yuanbyu replied:
This is for notation. We make it clear it is optional when talking about input `P`.
